sky/serve/autoscalers.py

```diff
@@ -411,54 +290,213 @@ def evaluate_scaling(
                     # and restart.
                     return []
 
-        self._set_target_num_replica_with_hysteresis()
+        scaling_options = []
 
-        scaling_options: List[AutoscalerDecision] = []
-        all_replica_ids_to_scale_down: List[int] = []
+        # If rolling update is in progress, we scale down old replicas based on
+        # the number of ready new replicas and the traffic is directed to both
+        # old and new replicas. Or, for blue_green update, once there is
+        # min_replicas number of ready new replicas, we will direct all traffic
+        # to them, we can scale down all old replicas.
+        # TODO(MaoZiming,zhwu): corner case: We should make sure the fallback
+        # replicas are ready before scaling down the old replicas to avoid the
+        # situation that all the ready new replicas are preempted together.
+        scaling_options.extend(
+            _to_scale_down_options(
+                self._select_outdated_replicas_to_scale_down(
+                    replica_infos, active_versions)))
+
+        # If the latest version is ever ready, we can proceed to evaluate from
+        # the implementations in subclasses.
+        scaling_options.extend(self._evaluate_scaling(replica_infos))
+
+        if not scaling_options:
+            logger.info('No scaling needed.')
+
+        return scaling_options
 
-        # Case 1. If rolling update is in progress, we scale down old replicas
-        # based on the number of ready new replicas and the traffic is directed
-        # to both old and new replicas.
-        # Or, for blue_green update, once there is min_replicas number of ready
-        # new replicas, we will direct all traffic to them, we can scale down
-        # all old replicas.
-        all_replica_ids_to_scale_down.extend(
-            self.select_outdated_replicas_to_scale_down(replica_infos))
 
-        # Case 2. when latest_nonterminal_replicas is less
+class HysteresisAutoscaler(Autoscaler):
```

Comment:
It is a bit weird calling it `HystersisAutoscaler`, as Hysteresis is just a technique to avoid keeping scaling up and down.

Also, in what case would this autoscaler ever scale? Is it just when a user updates the number of desired replicas? Or, is this just an internal utility class?
If it is a utility class, it would be good to make it clear in the comment and hide the class by pretending `_` to the name.

Reply:
Good point! It is indeed an utility class. Changed name to `_AutoscalerWithHysteresis`. I also change the format from `xxxAutoscaler` to `AutoscalerWithxxx`. PTAL!